# Daily Challenge

## What You Will Create
Implementation of Min-Max scaling and Z-score normalization.
Steps for data merging and dimensionality reduction using PCA.
Analysis of detected outliers, including visualizations.
Results of PCA, including the transformed dataset.

In [1]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler, StandardScaler
from scipy import stats
from scipy.stats import zscore
from scipy.stats.mstats import winsorize

In [4]:
col_names = ['age', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country', 'income']

adult = pd.read_csv('adult.data', names=col_names, na_values=' ?', skipinitialspace=True)

adult.info()
adult.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   
5   37           Private  284582    Masters             14   
6   49           Private  160187        9th              5   
7   52  Self-emp-not-inc  209642    HS-grad              9   
8   31           Private   45781    Masters             14   
9   42           Private  159449  Bachelors             13   

          marital-status         occupation   relationship   race     sex  \
0          Never-married       Adm-clerical  Not-in-family  White    Male   
1     Married-civ-spouse    Exec-managerial        Husband  White    Male   
2               Divorced  Handlers-cleaners  Not-in-family  White    Male   
3     Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4     Married-civ-spouse     Prof-specialty           Wife  Black  Female   
5     Married-civ-spouse    Exec-managerial           Wife  White  Female   
6  Married-spouse-absent      Other-service  Not-in-family  Black  Female   
7     Married-civ-spouse    Exec-managerial        Husband  White    Male   
8          Never-married     Prof-specialty  Not-in-family  White  Female   
9     Married-civ-spouse    Exec-managerial        Husband  White    Male   

   capital-gain  capital-loss  hours-per-week native-country income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K  
5             0             0              40  United-States  <=50K  
6             0             0              16        Jamaica  <=50K  
7             0             0              45  United-States   >50K  
8         14084             0              50  United-States   >50K  
9          5178             0              40  United-States   >50K

In [5]:
# checking for duplicate strings
adult.duplicated().sum()

24

In [6]:
# obtaining general information about numeric columns
adult.describe()

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

<div style="border:solid green 2px; padding: 20px">
    
- The dataset contains 32561 rows and 15 columns. There are no missing values.
- All columns in their respective types.
- There are 24 duplicates in the dataset and they need to be deleted.

In [7]:
# duplicate removal
adult = adult.drop_duplicates()
len(adult)

32537

## Apply Min-Max scaling and Z-score normalization to appropriate numerical features in the dataset

In [8]:
# Apply Min-Max scaling
min_max_scaler = MinMaxScaler()
df_min_max_scaled = pd.DataFrame(min_max_scaler.fit_transform(adult[['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss', 'hours-per-week']]), columns=['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss', 'hours-per-week'])

print("Min-Max Scaled Dataset:")
df_min_max_scaled

Min-Max Scaled Dataset:


age    fnlwgt  education-num  capital-gain  capital-loss  \
0      0.301370  0.044302       0.800000      0.021740           0.0   
1      0.452055  0.048238       0.800000      0.000000           0.0   
2      0.287671  0.138113       0.533333      0.000000           0.0   
3      0.493151  0.151068       0.400000      0.000000           0.0   
4      0.150685  0.221488       0.800000      0.000000           0.0   
...         ...       ...            ...           ...           ...   
32532  0.136986  0.166404       0.733333      0.000000           0.0   
32533  0.315068  0.096500       0.533333      0.000000           0.0   
32534  0.561644  0.094827       0.533333      0.000000           0.0   
32535  0.068493  0.128499       0.533333      0.000000           0.0   
32536  0.479452  0.187203       0.533333      0.150242           0.0   

       hours-per-week  
0            0.397959  
1            0.122449  
2            0.397959  
3            0.397959  
4            0.397959  
...               ...  
32532        0.377551  
32533        0.397959  
32534        0.397959  
32535        0.193878  
32536        0.397959  

[32537 rows x 6 columns]

In [9]:
# Apply Z-score normalization
z_score_scaler = StandardScaler()
df_z_score_normalized = pd.DataFrame(z_score_scaler.fit_transform(adult[['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss', 'hours-per-week']]), columns=['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss', 'hours-per-week'])

print("Z-Score Normalized Dataset:")
df_z_score_normalized

Z-Score Normalized Dataset:


age    fnlwgt  education-num  capital-gain  capital-loss  \
0      0.030390 -1.063569       1.134777      0.148292     -0.216743   
1      0.836973 -1.008668       1.134777     -0.145975     -0.216743   
2     -0.042936  0.245040      -0.420679     -0.145975     -0.216743   
3      1.056950  0.425752      -1.198407     -0.145975     -0.216743   
4     -0.776193  1.408066       1.134777     -0.145975     -0.216743   
...         ...       ...            ...           ...           ...   
32532 -0.849519  0.639678       0.745913     -0.145975     -0.216743   
32533  0.103716 -0.335436      -0.420679     -0.145975     -0.216743   
32534  1.423579 -0.358779      -0.420679     -0.145975     -0.216743   
32535 -1.216148  0.110930      -0.420679     -0.145975     -0.216743   
32536  0.983625  0.929812      -0.420679      1.887635     -0.216743   

       hours-per-week  
0           -0.035664  
1           -2.222483  
2           -0.035664  
3           -0.035664  
4           -0.035664  
...               ...  
32532       -0.197650  
32533       -0.035664  
32534       -0.035664  
32535       -1.655530  
32536       -0.035664  

[32537 rows x 6 columns]

<div style="border:solid green 2px; padding: 20px">
In this dataset, MinMaxScaler is used for Min-Max scaling, and StandardScaler is used for Z-score normalization. The resulting dataframes (df_min_max_scaled and df_z_score_normalized) will contain the normalized values for the specified numerical features.

## Identify any outliers in numerical features using statistical methods (like Z-score, IQR)

In [17]:
# Identify data points with Z-scores higher than 3 or lower than -3
z_score_outliers = df_z_score_normalized[(df_z_score_normalized > 3) | (df_z_score_normalized < -3)].dropna()

print("Identified outliers:")
z_score_outliers

Identified outliers:


Empty DataFrame
Columns: [age, fnlwgt, education-num, capital-gain, capital-loss, hours-per-week]
Index: []

In [16]:
# Identify outliers using IQR
Q1 = adult.quantile(0.25)
Q3 = adult.quantile(0.75)
IQR = Q3 - Q1
outliers_iqr = ((adult < (Q1 - 1.5 * IQR)) | (adult > (Q3 + 1.5 * IQR))).any(axis=1)

print("Outliers identified using IQR:")
adult[outliers_iqr]

C:\Users\Acer\AppData\Local\Temp\ipykernel_6324\1580326267.py:5: FutureWarning: Automatic reindexing on DataFrame vs Series comparisons is deprecated and will raise ValueError in a future version. Do `left, right = left.align(right, axis=1, copy=False)` before e.g. `left == right`
  outliers_iqr = ((adult < (Q1 - 1.5 * IQR)) | (adult > (Q3 + 1.5 * IQR))).any(axis=1)


Outliers identified using IQR:


age         workclass  fnlwgt    education  education-num  \
0       39         State-gov   77516    Bachelors             13   
1       50  Self-emp-not-inc   83311    Bachelors             13   
6       49           Private  160187          9th              5   
8       31           Private   45781      Masters             14   
9       42           Private  159449    Bachelors             13   
...    ...               ...     ...          ...            ...   
32545   39         Local-gov  111499   Assoc-acdm             12   
32548   65  Self-emp-not-inc   99359  Prof-school             15   
32553   32           Private  116138      Masters             14   
32559   22           Private  201490      HS-grad              9   
32560   52      Self-emp-inc  287927      HS-grad              9   

              marital-status       occupation   relationship  \
0              Never-married     Adm-clerical  Not-in-family   
1         Married-civ-spouse  Exec-managerial        Husband   
6      Married-spouse-absent    Other-service  Not-in-family   
8              Never-married   Prof-specialty  Not-in-family   
9         Married-civ-spouse  Exec-managerial        Husband   
...                      ...              ...            ...   
32545     Married-civ-spouse     Adm-clerical           Wife   
32548          Never-married   Prof-specialty  Not-in-family   
32553          Never-married     Tech-support  Not-in-family   
32559          Never-married     Adm-clerical      Own-child   
32560     Married-civ-spouse  Exec-managerial           Wife   

                     race     sex  capital-gain  capital-loss  hours-per-week  \
0                   White    Male          2174             0              40   
1                   White    Male             0             0              13   
6                   Black  Female             0             0              16   
8                   White  Female         14084             0              50   
9                   White    Male          5178             0              40   
...                   ...     ...           ...           ...             ...   
32545               White  Female             0             0              20   
32548               White    Male          1086             0              60   
32553  Asian-Pac-Islander    Male             0             0              11   
32559               White    Male             0             0              20   
32560               White  Female         15024             0              40   

      native-country income  
0      United-States  <=50K  
1      United-States  <=50K  
6            Jamaica  <=50K  
8      United-States   >50K  
9      United-States   >50K  
...              ...    ...  
32545  United-States   >50K  
32548  United-States  <=50K  
32553         Taiwan  <=50K  
32559  United-States  <=50K  
32560  United-States   >50K  

[13554 rows x 15 columns]

<div style="border:solid green 2px; padding: 20px">
No outliers (greater than 3 and less than -3) were detected using the Z-scores method. When using IQR, 13554 lines with outliers were detected.

## Choose an appropriate method to handle these outliers (e.g., trimming, capping, transformation) and apply it to the dataset

In [19]:
# Winsorize the numerical columns to handle outliers
df_winsorized = adult.apply(lambda x: winsorize(x, limits=[0.05, 0.05]))

# Convert the winsorized array back to a DataFrame
df_winsorized = pd.DataFrame(df_winsorized, columns=adult.columns)

print("Winsorized Dataset:")
df_winsorized

Winsorized Dataset:


age         workclass  fnlwgt   education  education-num  \
0       39  Self-emp-not-inc   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race     sex  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty      Unmarried  Black  Female   
...                   ...                ...            ...    ...     ...   
32556  Married-civ-spouse       Tech-support      Unmarried  White  Female   
32557  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558           Separated       Adm-clerical      Unmarried  White  Female   
32559       Never-married       Adm-clerical      Own-child  White    Male   
32560  Married-civ-spouse    Exec-managerial      Unmarried  White  Female   

       capital-gain  capital-loss  hours-per-week native-country income  
0              2174             0              40  United-States  <=50K  
1                 0             0              18  United-States  <=50K  
2                 0             0              40  United-States  <=50K  
3                 0             0              40  United-States  <=50K  
4                 0             0              40          India  <=50K  
...             ...           ...             ...            ...    ...  
32556             0             0              38  United-States  <=50K  
32557             0             0              40  United-States   >50K  
32558             0             0              40  United-States  <=50K  
32559             0             0              20  United-States  <=50K  
32560          5013             0              40  United-States   >50K  

[32537 rows x 15 columns]

<div style="border:solid green 2px; padding: 20px">
Handling outliers in a dataset can be approached in various ways, and the choice depends on the nature of data and the goals of analysis. 
Capping (Winsorizing) involves setting a threshold beyond which values are capped at a certain value. This helps in preventing extreme values from affecting the analysis. In this dataset, the winsorize function is applied to each column of the dataset, capping the values at the 5th and 95th percentiles. The effect of winsorizing is to reduce the impact of extreme values on the dataset.